tsl/src/gapfill/exec.c

```diff
@@ -178,6 +189,182 @@ gapfill_state_create(CustomScan *cscan)
 	return (Node *) state;
 }
 
+/*
+ * lookup operator name in pg_operator
+ *
+ * returns empty string for unknown operator
+ */
+static char *
+get_oprname(Oid oprid)
+{
+	HeapTuple	opertup;
+	char	   *result = "";
+
+	if (!OidIsValid(oprid))
+		return result;
+
+	opertup = SearchSysCache1(OPEROID, ObjectIdGetDatum(oprid));
+	if (HeapTupleIsValid(opertup))
+	{
+		Form_pg_operator operform = (Form_pg_operator) GETSTRUCT(opertup);
+
+		result = pstrdup(NameStr(operform->oprname));
+
+		ReleaseSysCache(opertup);
+	}
+
+	return result;
+}
+
+/*
+ * lookup cast func oid in pg_cast
+ *
+ * throws an error if no cast can be found
+ */
+static Oid
+get_castfunc(Oid source, Oid target)
+{
+	Oid			result = InvalidOid;
+	HeapTuple	casttup;
+
+	casttup = SearchSysCache2(CASTSOURCETARGET, ObjectIdGetDatum(source), ObjectIdGetDatum(target));
+	if (HeapTupleIsValid(casttup))
+	{
+		Form_pg_cast castform = (Form_pg_cast) GETSTRUCT(casttup);
+
+		result = castform->castfunc;
+		ReleaseSysCache(casttup);
+	}
+
+	if (!OidIsValid(result))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("error looking up cast from %d to %d", source, target)));
+
+	return result;
+}
+
+/*
+ * returns true if v1 and v2 reference the same object
+ */
+static bool
+var_equal(Var *v1, Var *v2)
+{
+	return v1->varno == v2->varno && v1->varattno == v2->varattno && v1->vartype == v2->vartype;
+}
+
+static void
+deduce_gapfill_boundary(GapFillState *state, GapFillBoundary boundary, int64 *value)
+{
+	CustomScan *cscan = castNode(CustomScan, state->csstate.ss.ps.plan);
+	FuncExpr   *func = linitial(cscan->custom_private);
+	FromExpr   *jt = lfourth(cscan->custom_private);
+	List	   *vars;
+	ListCell   *lc;
+
+	const char *op1,
+			   *op2;
+
+	if (GAPFILL_START == boundary)
+	{
+		op1 = ">";
+		op2 = ">=";
+	}
+	else
+	{
+		op1 = "<";
+		op2 = "<=";
+	}
+
+	/*
+	 * if the second argument to time_bucket_gapfill is not a column reference
+	 * we cannot match WHERE clause to the time column, we don't check for Var
+	 * directly though cause it might be wrapped in a cast
+	 */
+	vars = pull_var_clause(lsecond(func->args), 0);
+	if (list_length(vars) != 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid ts: need to refer to column if no start or finish is supplied")));
```

Comment:
This error message seems like it needs more information. Possibly refer the `gapfill` directly?

Reply:
added an errorhint where appropriate and also added function name to error messages